### PR TITLE
Allow security_rule objects to have rule IDs different from the object IDs

### PR DESCRIPTION
**Pull request by xcrzx**
## What does this PR do?

Change validation check that enforces saved object ID to match `rule_id` for security solution prebuilt detection rules. Allow saved object IDs to have arbitrary suffixes.

## Why is it important?

Security Solution migrates the detection rules package from storing a single saved object per rule to multiple saved objects. See https://github.com/elastic/kibana/issues/137420 for more context regarding the change.

The package will contain rule saved objects with rule id and version in the name (`security_rule/[ruleId]_[ruleVersion].json`) with the following content:

```json
{
  "id": "[ruleId]_[ruleVersion]",
  "type": "security-rule",
  "attributes": {
    "rule_id": "[ruleId]",
    "version": "[ruleVersion]",
    // Other rule attributes
  }
}
```

So the saved object ID and the rule ID do not match anymore, making this validation check always return an error.

## Checklist

- [x] I have added test packages to [`test/packages`](https://github.com/elastic/package-spec/tree/main/test/packages) that prove my change is effective.
- [x] I have added an entry in [`spec/changelog.yml`](https://github.com/elastic/package-spec/blob/main/spec/changelog.yml).

## Related issues

- Closes https://github.com/elastic/package-spec/issues/459

### CODE REVIEW

File: test/packages/good/kibana/security_rule/000047bb-b27a-47ec-8b62-ef1a5d2c9e19_5.json [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      "Okta",
+      "Continuous Monitoring",
+      "SecOps",
+      "Identity and Access"
+    ],
+    "timestamp_override": "event.ingested",
+    "type": "query",
+    "version": 5
+  },
+  "id": "000047bb-b27a-47ec-8b62-ef1a5d2c9e19_5",
```
Comment (banderror) [inline test/packages/good/kibana/security_rule/000047bb-b27a-47ec-8b62-ef1a5d2c9e19_5.json:34]: "id": "000047bb-b27a-47ec-8b62-ef1a5d2c9e19"?

File: code/go/internal/validator/semantic/validate_kibana_matching_object_ids.go [hunk 2]
```diff
@@ -41,16 +41,11 @@ func ValidateKibanaObjectIDs(fsys fspath.FS) ve.ValidationErrors {
...
+			_, err := objectFile.Values("$.attributes.rule_id")
 			if err != nil {
 				errs = append(errs, errors.Wrapf(err, "unable to get rule ID in file [%s]", fsys.Path(filePath)))
 				continue
 			}
-
-			if ruleID != objectID {
-				errs = append(errs, errors.New("rule ID is different from the object ID"))
-				continue
-			}
```
[the remark this answers is not in the document]
Comment (xcrzx) [reply code/go/internal/validator/semantic/validate_kibana_matching_object_ids.go]: > Could that (at least partially) address the possibility of someone copy-pasting a rule asset, updating the object ID but forgetting to update the `rule_id`?

Good idea 👍 Changed the validation for checking if the object ID starts with the rule ID for a given rule.

> Another question: is there an explicit validation somewhere that verifies the uniqueness of saved object IDs? Not the file names, but the combinations of SO id and type from the file contents?

I don't know, haven't seen checks like that in this repo. It looks like only filename uniqueness enforces saved object id uniqueness.

**Review by banderror — DISMISSED** [review]
LGTM!

File: code/go/internal/validator/semantic/validate_kibana_matching_object_ids.go [hunk 3]
```diff
@@ -47,8 +47,9 @@ func ValidateKibanaObjectIDs(fsys fspath.FS) ve.ValidationErrors {
 				continue
 			}
 
-			if ruleID != objectID {
-				errs = append(errs, errors.New("rule ID is different from the object ID"))
+			if !strings.HasPrefix(objectID.(string), ruleID.(string)) {
```
Comment (mrodm) [inline code/go/internal/validator/semantic/validate_kibana_matching_object_ids.go:50]: These type assertions could result in a panic if `objectId` or `ruleID` are not string.
Could you add those checks before this if statement ? Something like:
```suggestion
            stringObjectID, ok := objectID.(string)
            if !ok {
                errs = append(errs, errors.Wrapf(err, ...)
                continue
            }
            stringRuleID, ok := ruleID.(string)
            if !ok {
                errs = append(errs, errors.Wrapf(err, ...)
                continue
            }
			if !strings.HasPrefix(stringObjectID, stringRuleID) {
```

**Review by mrodm — COMMENTED** [review]
LGTM, just a note about avoiding possible panics in some type assertions.

File: code/go/internal/validator/semantic/validate_kibana_matching_object_ids.go (hunk 3 above)
Comment (xcrzx) [reply code/go/internal/validator/semantic/validate_kibana_matching_object_ids.go]: @mrodm Thanks for the suggestion 👍 Added additional checks

**Review by mrodm — APPROVED** [review]
Thanks!

LGTM! :+1:

File: code/go/internal/validator/semantic/validate_kibana_matching_object_ids.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (spong) [reply code/go/internal/validator/semantic/validate_kibana_matching_object_ids.go]: > >Another question: is there an explicit validation somewhere that verifies the uniqueness of saved object IDs? Not the file names, but the combinations of SO id and type from the file contents?
>
> I don't know, haven't seen checks like that in this repo. It looks like only filename uniqueness enforces saved object id uniqueness.

As far as I understand it's indeed the filename uniqueness that's enforcing the SO id uniqueness as well. Current behavior right now is last-one-in-wins if there are duplicate SO ID's ([comment](https://github.com/elastic/kibana/issues/128202#issuecomment-1084600441)).

**Comment by xcrzx** [comment]
@mrodm It seems like this PR has all the required approves. How could I merge it? The merge button is inactive, saying, `You're not authorized to merge this pull request.`

**Comment by mrodm** [comment]
> @mrodm It seems like this PR has all the required approves. How could I merge it? The merge button is inactive, saying, `You're not authorized to merge this pull request.`

@xcrzx As all the required approves are there, I'll merge it.

**Pull request merged**